### PR TITLE
MANTA-3674 manta-adm should have a channel guard

### PR DESCRIPTION
**Pull request by timfoster**
**This PR was migrated from https://github.com/joyent/sdc-manta/pull/6**

MANTA-2769 manta-init should be channel aware
MANTA-4159 manta-init should have an option to specify an image version substring
MANTA-3674 manta-adm should have a channel guard

This PR was migrated-from-gerrit, <https://cr.joyent.us/#/c/6808/>.
The raw archive of this CR is [here](https://github.com/joyent/gerrit-migration/tree/master/archive/6808).
See [MANTA-4594](https://smartos.org/bugview/MANTA-4594) for info on Joyent Eng's migration from Gerrit.

##### @timfoster commented at 2019-08-19T15:38:33

> Patch Set 1:
> 
> New commits:  
>     commit 0f70ed94a320e8ccc869dfa16f0d6d4713b16b50  
>     MANTA-2769 manta-init should be channel aware

##### @timfoster commented at 2019-08-19T15:52:53

> Patch Set 2:
> 
> New commits:  
>     commit 4894f32348e2db81439970a97d8a3ba79db02fa3  
>     fix 'make check' errors

##### @timfoster commented at 2019-08-20T10:15:40

> Patch Set 3:
> 
> New commits:  
>     commit 199d8a9de142663146d544059e0abd77884f3118  
>     Switch to a default -B value of ''

##### @timfoster commented at 2019-09-18T11:16:05

> Patch Set 4:
> 
> New commits:  
>     commit 798ac0c8c86d7ee315fa25eea9f4e5ba75d65e4e  
>     MANTA-3674 manta-adm should have a channel guard
>     
>     commit a28814f2334a61607edac1b02bf03339b838490a  
>     MANTA-2769 manta-init should be channel aware

##### @timfoster commented at 2019-09-18T11:29:11

> Patch Set 4:
> 
> This needs a little more work - the changes were developed on two
> branches, and I now want to see if I can refactor the lookup of the SAPI
> 'channel' option so that there's less duplicated code.

##### @timfoster commented at 2019-09-19T08:28:29

> Patch Set 5:
> 
> New commits:  
>     commit 3e29ce504d3dd40b276c6346fa63a9066af8d3a1  
>     share SAPI channel lookup code between manta-init and manta-adm
>     
>     commit 024c59efc33fb7f01c3f4a195956dcc5fc3507b5  
>     MANTA-3674 manta-adm should have a channel guard
>     
>     commit b71fa427a1d19d5e2537d412f5474b43b4569cf7  
>     MANTA-2769 manta-init should be channel aware

##### @KodyKantor commented at 2019-10-03T15:20:40

> Patch Set 5:
> 
> (7 comments)
> 
> I think this looks really good. Well done!
> 
> ###### cmd/manta-adm.js#881  
> 
> I prefer the 'return' to be on a separate line. This repo is inconsistent about this sort of thing though so it's not a big deal.
> 
> ###### cmd/manta-adm.js#881  
> 
> No problem, I'll fix that.
> 
> ###### cmd/manta-adm.js#970  
> 
> Should these be spaces (based on the previous two flags)?
> 
> ###### cmd/manta-adm.js#970  
> 
> Yep, thanks
> 
> ###### cmd/manta-adm.js#973  
> 
> Are we missing a comma in this string? It reads a little funny to me.
> 
> ###### cmd/manta-adm.js#973  
> 
> ugh, yes that's terrible - I'll reword.
> 
> ###### cmd/manta-init.js#430  
> 
> Since the default value for ARGV.branch is already '', what do you think about removing this condition?
> 
> I like the ability to not search on a specific channel, but having a dual-purpose bool/string flag feels a little weird to me. Maybe we can limit that behavior to just the -C flag by eliminating the similar behavior here for -B?

##### @timfoster commented at 2019-10-03T16:22:41

> Patch Set 5:
> 
> (7 comments)
> 
> Thanks for the review! I'll post the follow-up changes if you can suggest the best approach for lib/adm.js line 4158 (I'm totally easy, either way)
> 
> ###### cmd/manta-init.js#430  
> 
> That's reasonable. Fwiw, earlier versions of this code had our default for -B the same as the current code, 'master', but Trent persuaded me out of doing that, hence this proposed change, but I've tested that the default of '' really does give us a string rather than a boolean, so we're cool.
> 
> ###### lib/adm.js#4158  
> 
> Is there an extra tab here?
> 
> ###### lib/adm.js#4158  
> 
> Yeah, I wasn't sure how to indent this code.
> 
> For editors where 1 tab == 4 spaces, indenting the 2nd and 3rd clauses of the for-loop by only 4 spaces will result in those lines appearing to start on the same column as the body of the loop.
> 
> So instead, I added an extra tab indent so that it was clear
> what was going on. I don't mind removing that extra indent if you prefer? So we have this:
> 
>     for (var i = 0;
>             i < this.ma_plan[svcname][cnid].length;
>             i++) {
>         content;
>         of;
>         loop;
>     }
> 
> rather than
> 
>     for (var i = 0;
>         i < this.ma_plan[svcname][cnid].length;
>         i++) {
>         content;
>         of;
>         loop;
>     }
> 
> I'm easy either way?
> 
> ###### lib/adm.js#4158  
> 
> I would prefer that this is converted into four spaces since it's a line continuation. When 'for' statements wrap lines sometimes people add an extra blank line before the loop body to differentiate the content a bit.
> 
> I agree that our mix tab and space style can be really frustrating.
> 
> ###### lib/adm.js#4158  
> 
> No problem, will do!
> 
> ###### lib/adm.js#4179  
> 
> Perhaps we need a tab here.
> 
> ###### lib/adm.js#4179  
> 
> yep, thanks!

##### @KodyKantor commented at 2019-10-04T14:23:38

> Patch Set 5:
> 
> (2 comments)
> 
> ###### lib/adm.js#4189  
> 
> Should this also be spaces, like the next line?
> 
> ###### lib/adm.js#4189  
> 
> Ugh, thanks - I'm basically just doing what jsstyle/jsl wants me to at this point. Solaris ON-style mixed-tab/space indents are nuts imho :-)

##### @timfoster commented at 2019-10-04T15:06:58

> Patch Set 6:
> 
> New commits:  
>     commit ed19d3ea2a951c0ae06534379d38e037d7ec2e17  
>     address kkantor's review feedback

##### @timfoster commented at 2019-10-04T15:07:15

> Patch Set 5:
> 
> (2 comments)
> 
> I've addressed these changes in patchset 6
> 
> ###### lib/adm.js#4189  
> 
> Yeah, I generally agree with your assessment of the style. Javascript's tendencies for very long lines doesn't help. Does jsstyle complain if these are spaces instead of tabs?
> 
> ###### lib/adm.js#4189  
> 
> nope, jsstyle is fine here

##### @KodyKantor commented at 2019-10-04T15:09:38

> Patch Set 6: Code-Review+1
### comments copied from PR 6

### CODE REVIEW

**Comment by timfoster** [comment]
I'm going to address the remaining review comments on #6 , but will push any updates to this branch.

File: lib/common.js [hunk 1]
```diff
@@ -937,6 +937,47 @@ function getServerNicTags(cb) {
...
+	};
+
+	sapi.listApplications(search, function (err, apps) {
+		if (err) {
+			log.error(err, 'failed to list applications');
+			return (cb(err));
+		}
+
+		log.debug('found sdc application');
+		assert.ok(apps.length <= 1);
```
Comment (trentm) [inline lib/common.js]: probably want to check that it is exactly 1. If it is zero, then your next `apps[0]` is going to break. However if it is 0 then no way you got to this point. ;)

Comment (timfoster) [reply lib/common.js]: Ha, good point. I'll change that just to make it very clear.

**Comment by timfoster** [comment]
Thanks for the review!

**Pull request merged**